core/internal/api/credentials.go

```diff
@@ -55,3 +65,237 @@ func (c *apiKeyCredentialProvider) Apply(req *http.Request) error {
 	)
 	return nil
 }
+
+var _ CredentialProvider = &oauth2CredentialProvider{}
+
+// OAuth2CredentialProvider creates a credentials provider that exchanges a JWT
+// for an access token via an authorization server. The access token is then used
+// to authenticate API requests.
+//
+// The JWT is supplied via a file path that is passed in as an environment
+// variable. When the OAuth2CredentialProvider is applied, it exchanges the JWT
+// for an access token. Once the token is received, it is saved to the
+// credentials file along with its expiration. The expiration is checked each
+// time the access token is used, and refreshed if it is at or near expiration.
+func NewOAuth2CredentialProvider(
+	settings *settings.Settings,
```

Comment:
what do you think about more explicit about the arguments here and using settings in the caller site?

Reply:
thinking the same for the apikey one as well... i know it modified a bit the general newprovider, but seems cleaner that way if we don't mix the settings with the providers

Reply:
That makes sense to me. We could pass settings into NewCredentialProvider, but NewAPIKeyCredentialProvider and NewOAuth2CredentialProvider could be more explicit about what settings they use.

Reply:
yep that's exactly what i was thinking :)

Reply:
Updated. I didn't change NewAPIKeyCredentialProvider though because it calls EnsureApiKey on the settings object and didn't want to refactor that within this PR